dogechain-lab/dogechain
# [Txpool] Clip most enqueued length account when memory usage high

# Description

Some users might keep sending out huge amount future transactions, which might blocked the whole network. Nodes could not recover to normal unless it prune stale accounts.

This PR adds a supplementary strategy to this use case.

It only clips all transactions of the account who enqueued most, and the clipping checking is periodically.

So it is OK to add more transactions to the pool even it reaches the threshold. Honest users will not be bothered, even if the nonce hole exists due to network reasons. They can re-send the missing nonce transactions, replace those future transactions, or just wait for pruning when outdated.

The `RPC` or `validator` nodes could adjust `clipping` flags more reasonable, the default flags are most suitable for most use cases.

# Changes include


# Checklist


## Testing


### Manual tests

* Set up a dev node, using `--max-slots 300 --clipping.tick-seconds 1 --clipping.memory-threshold 1` flags.
* Subscribe the txpool `pruned-enqueued` events.
* Use `Metamask` to send 3 transactions, 1 from account1, 2 from account2.
* Reset `Metamask` accounts to stop re-sending over and over again.

In the PR branch, you should see the pruning of account2 transactions, and `txpool_content` would get only account1's transactions.

# Documentation update

Will update the documentation when it is merged.

# Additional comments

The PR is inspired by the 0xPolygon Edge project PR [689](https://github.com/0xPolygon/polygon-edge/pull/689), [690](https://github.com/0xPolygon/polygon-edge/pull/690).

But we make it more sensible, because you should not punish everyone when there are only few guilty. `prune` and `reject` more transactions when memory high in `addTx` sounds buggy. e.g. What if the new transaction is the 'nonce hole patcher'?

@zivkovicmilos Hi, zivkov. You might be interested in this.
